### ISSUE TITLE
Only one screen is displayed

### ISSUE THREAD

**Issue by varac** [b12f/wlstreamer#2]
I start wlstreamer with `wlstreamer -o /dev/video2` and the go to the [Mozilla webrtc gum test](https://mozilla.github.io/webrtc-landing/gum_test.html). Only my external monitor is shown, when I move my focus to my laptop display I see in the wlstreamer output that it switched from `Recording DP-3` to `Recording LVDS-1`. However, the shared screencast in the browser still shows the same (unfocussed) screen as before.

```
❯ wlstreamer -o /dev/video2                                                             
Found screens: ["LVDS-1", "DP-3"]
Outputting to --file=/dev/video2
Switched focus
Found screens: ["DP-3", "LVDS-1"]
Outputting to --file=/dev/video2
Recording DP-3
Switched focus
Found screens: ["DP-3", "LVDS-1"]
Switched focus
Found screens: ["DP-3", "LVDS-1"]
Switched focus
Found screens: ["DP-3", "LVDS-1"]
Switched focus
Found screens: ["DP-3", "LVDS-1"]
Switched focus
Found screens: ["DP-3", "LVDS-1"]
Switched focus
Found screens: ["LVDS-1", "DP-3"]
Outputting to --file=/dev/video2
Recording LVDS-1
Switched focus
Found screens: ["DP-3", "LVDS-1"]
Outputting to --file=/dev/video2
Recording DP-3
Switched focus
Found screens: ["LVDS-1", "DP-3"]
Outputting to --file=/dev/video2
Recording LVDS-1
Switched focus
Found screens: ["DP-3", "LVDS-1"]
Outputting to --file=/dev/video2
Recording DP-3
Switched focus
Found screens: ["LVDS-1", "DP-3"]
Outputting to --file=/dev/video2
Recording LVDS-1
```

**Comment by varac** [comment]
When I watch the stream with `vlc v4l2:///dev/video2` sth weird happens:
First, I see a completely garbeled stream from my Laptop screen, when I started wlstreamer. When I focus my external monitor the stream looks fine. When I focus back on my Laptop screen the monitor screen still shows up in VLC.

**Comment by b12f** [comment]
Thanks for the report, I'm unsure what the problem might be.

* Which browser did you try the gUM test with? Chrome-based browsers might need `exclusive_caps=1` when enabling v4l2loopback
* Do both monitors have the same resolution? I've never tried sending different resolutions through loopback devices and am unsure how programs would handle it. If so, a good way to test this, would be to start wlstreamer once on either screen (perhaps the starting monitor determines the stream resolution?), and seeing if the other screen is garbled.

I'll see if I can get a `--verbose` option into wlstreamer somehow

**Comment by b12f** [comment]
After testing, it seems like this may be a problem with differing resolutions. I'm unsure how to fix this, let me do some searching.

**Comment by varac** [comment]
I already started the v4l2loopback kernel module with `modprobe v4l2loopback exclusive_caps=1 card_label=WfRecorder`
I indeed have different screen resolution:

```
❯ swaymsg -t get_outputs
Output LVDS-1 'Unknown 0x03A3 0x00000000'
  Current mode: 1366x768 @ 60.019001 Hz
  Position: 0,0
  Scale factor: 1.000000
  Scale filter: nearest
  Subpixel hinting: rgb
  Transform: normal
  Workspace: 9
  Max render time: off
  Adaptive sync: disabled
  Available modes:
    1366x768 @ 60.019001 Hz

Output DP-3 'Samsung Electric Company S24C650 0x00000000' (focused)
  Current mode: 1920x1080 @ 60.000000 Hz
  Position: 1366,0
  Scale factor: 1.000000
  Scale filter: nearest
  Subpixel hinting: unknown
  Transform: normal
  Workspace: 4 🌐
  Max render time: off
  Adaptive sync: disabled
  Available modes:
    720x400 @ 70.082001 Hz
    640x480 @ 59.939999 Hz
    640x480 @ 60.000000 Hz
    640x480 @ 66.667000 Hz
    640x480 @ 72.808998 Hz
    640x480 @ 75.000000 Hz
    720x480 @ 59.939999 Hz
    720x480 @ 59.939999 Hz
    720x480 @ 60.000000 Hz
    720x480 @ 60.000000 Hz
    720x576 @ 50.000000 Hz
    720x576 @ 50.000000 Hz
    800x600 @ 56.250000 Hz
    800x600 @ 60.317001 Hz
    800x600 @ 72.188004 Hz
    800x600 @ 75.000000 Hz
    832x624 @ 74.551003 Hz
    1024x768 @ 60.004002 Hz
    1024x768 @ 70.069000 Hz
    1024x768 @ 75.028999 Hz
    1280x720 @ 50.000000 Hz
    1280x720 @ 59.939999 Hz
    1280x720 @ 60.000000 Hz
    1280x720 @ 60.000000 Hz
    1152x864 @ 75.000000 Hz
    1280x800 @ 59.810001 Hz
    1440x900 @ 59.887001 Hz
    1280x1024 @ 60.020000 Hz
    1280x1024 @ 75.025002 Hz
    1600x900 @ 60.000000 Hz
    1680x1050 @ 59.953999 Hz
    1920x1080 @ 50.000000 Hz
    1920x1080 @ 59.939999 Hz
    1920x1080 @ 60.000000 Hz
    2179x2179 @ 35.535999 Hz
    1920x1080 @ 60.000000 Hz
```


And was trying with both firefox (`82.0.3+build1-0ubuntu0.20.10.1`) and chromium (`86.0.4240.198` and `88.0.4315.5`), both with the same result.

**Comment by b12f** [comment]
So I've tried playing around with `v4l2loopback-ctl set-caps` - setting a new stream resolution between recording one screen and the next. However, this doesn't seem to work on my machine. The `v4l2loopback` device does not change the advertised format when the input stream changes. Not sure if this is by design or if I'm doing it wrong.

**Comment by b12f** [comment]
I've asked for help: https://unix.stackexchange.com/questions/620263/how-do-i-change-the-resolution-capabilities-of-a-v4l2loopback-device-on-the-fly

However, an alternative would be to take the largest height and largest width before starting the stream (so, for example, a 1920x1080 and a 1600x1200 screen would result in a 1920x1200 stream), and then padding the sides with a black border. That way the resolution doesn't have to change on the fly. Would this be acceptable behaviour to you @varac?

**Comment by varac** [comment]
> However, an alternative would be to take the largest height and largest width before starting the stream (so, for example, a 1920x1080 and a 1600x1200 screen would result in a 1920x1200 stream), and then padding the sides with a black border. That way the resolution doesn't have to change on the fly. Would this be acceptable behaviour to you @varac?

Sure, sounds good as long it makes it stable.

**Comment by varac** [comment]
After reading yhttps://github.com/ammen99/wf-recorder/issues/112 I started to use the `sharescreen.sh` script (see [this comment](https://github.com/ammen99/wf-recorder/issues/112#issuecomment-729846219)  and it worked nicely for me.
wlstreamer again displayed a garbeled screen.
I investigated which wf-recorder parameters were different and only found that the `sharescreen.sh` script  was using `yuyv422` as output pixel format (good), whereas wlstreamer uses `yuv420p` (garbled). 
As I don't know anything about this difference I'll leave the interpretation up to you...

**Comment by varac** [comment]
I can confirm that just by replacing both occurrences of `yuv420p` by `yuyv422` in wlstreamer/src/main.rs produces the same stable quality as the sharescreen.sh script for me (only tested with laptop only, not with an external monitor).

**Comment by b12f** [comment]
OH cool. I tried that and it didn't seem to fix the issue for me. Can you open a PR, maybe I did something wrong. Thanks for the debugging!

**Comment by varac** [comment]
Testing with an external monitor I still have a garbled output, so I'd rather hold on merging https://github.com/b12f/wlstreamer/pull/4.

Another thing I noticed that when using `--verbose`, I see that wf-recorder asks for the output no. (and doesn't get an input ?):

```
❯ wlstreamer --verbose -o /dev/video0
Found recordable screens: ["LVDS-1", "HDMI-A-1"]
Outputting to /dev/video0
Please select an output from the list to capture (enter output no.):
1. Name: LVDS-1 Description: Unknown 0x03A3 0x00000000 (LVDS-1)
2. Name: HDMI-A-1 Description: Acer Technologies V223W LBQ0C09640A5 (HDMI-A-1)
Enter output no.:selected region 0 0 0 0
Choosing pixel format yuyv422
Output #0, video4linux2,v4l2, to '/dev/video0':
    Stream #0:0: Unknown: none (rawvideo)
[video4linux2,v4l2 @ 0x7f8120000cc0] Using AVStream.codec to pass codec parameters to muxers is deprecated, use AVStream.codecpar instead.
[swscaler @ 0x7f8120002580] Warning: data is not aligned! This can lead to a speed loss
Switched focus
Found recordable screens: ["HDMI-A-1", "LVDS-1"]
Outputting to /dev/video0
Recording HDMI-A-1
Please select an output from the list to capture (enter output no.):
1. Name: LVDS-1 Description: Unknown 0x03A3 0x00000000 (LVDS-1)
2. Name: HDMI-A-1 Description: Acer Technologies V223W LBQ0C09640A5 (HDMI-A-1)
Enter output no.:selected region 0 0 0 0
Choosing pixel format yuyv422
Output #0, video4linux2,v4l2, to '/dev/video0':
    Stream #0:0: Unknown: none (rawvideo)
[video4linux2,v4l2 @ 0x7fa28c000cc0] Using AVStream.codec to pass codec parameters to muxers is deprecated, use AVStream.codecpar instead.
Switched focus
Found recordable screens: ["LVDS-1", "HDMI-A-1"]
Outputting to /dev/video0
Recording LVDS-1
Switched focus
Found recordable screens: ["LVDS-1", "HDMI-A-1"]
Please select an output from the list to capture (enter output no.):
1. Name: LVDS-1 Description: Unknown 0x03A3 0x00000000 (LVDS-1)
2. Name: HDMI-A-1 Description: Acer Technologies V223W LBQ0C09640A5 (HDMI-A-1)
Enter output no.:selected region 0 0 0 0
Choosing pixel format yuyv422
Output #0, video4linux2,v4l2, to '/dev/video0':
    Stream #0:0: Unknown: none (rawvideo)
[video4linux2,v4l2 @ 0x7f387c000cc0] Using AVStream.codec to pass codec parameters to muxers is deprecated, use AVStream.codecpar instead.
[swscaler @ 0x7f387c002580] Warning: data is not aligned! This can lead to a speed loss
Switched focus
Found recordable screens: ["LVDS-1", "HDMI-A-1"]
Switched focus
...
```

**Comment by b12f** [comment]
It gets an input, I write the screen index to the process stdin on line 92

**Comment by b12f** [comment]
I've changed that around, where it now sends the screen to record via the `-o` option. I guess I missed that when first using wf-recorder.

**Comment by varac** [comment]
I captured the wlstreamer stream with cheese to illustrate what I encounter when focussing on external monitor and back to my Laptop: [2020-11-18-213531.zip](https://github.com/b12f/wlstreamer/files/5562600/2020-11-18-213531.zip)

**Comment by b12f** [comment]
Yeah for me it looks a tad different but basically the same behaviour; I'm pretty sure it's due to the different resolutions.

I really hope it's possible to broadcast/force a different resolution via the loopback device, since doing the black border thing could turn out pretty ugly code wise.

The idea would be to have a different loopback device for each screen resolution, and then one which will be your output. ffmpeg then takes one of the screen outputs, and puts a black border around this, after which it feeds it into the output loopback. You'd need `numberOfDifferentResolutions + 1` loopback devices to be available for this to work, unless something like https://github.com/schauveau/wf-recorder-x gets merged into wf-recorder, and we can set ffmpeg filters directly on the wf-recorder options. The latter seems to not be happening any time soon though.

I'm personally too inexperienced with v4l2loopback and video encoding in general to know how to make a loopback device change resolutions and put a PR out to the v4l2 guy.

**Comment by b12f** [comment]
If you can, could you force the same resolution via `sway output $name mode $mode` on both screens, and after a `rmmod && modprobe` see if it works fine?

**Comment by varac** [comment]
> If you can, could you force the same resolution via sway output $name mode $mode on both screens, and after a rmmod && modprobe see if it works fine?

I can't :crying_cat_face:  my laptop has 1366x768 as only resolution, and my external monitor doesn't support this resolution (see 
https://github.com/b12f/wlstreamer/issues/2#issuecomment-729628146)

**Comment by varac** [comment]
But I try with another external display hopefully tomorrow.

**Comment by varac** [comment]
> unless something like https://github.com/schauveau/wf-recorder-x gets merged into wf-recorder

What's your plan regarding wf-recorder-x ? It would be really nice if both project can get merged.

**Comment by b12f** [comment]
No plans, it seems to be rather hard to get it merged: https://github.com/ammen99/wf-recorder/issues/38

**Comment by b12f** [comment]
@varac try running current master. The CLI options have changed a bit and you'll need to have more loopback devices available, but I'm thinking it should run. It's pretty hacky, but most of the hackiness is bypassed when all screens share the same resolution ;)

**Comment by varac** [comment]
Awesome! Works like a charm :partying_face: 
At least with the [mozilla webrtc gum page](https://mozilla.github.io/webrtc-landing/gum_test.html), need to carry on testing in real life (zoom, jitsi etc).

Ftr, it should get documented how to start multiple loopback devices, i.e. I used (for my laptop display and one external monitor):

    sudo modprobe v4l2loopback exclusive_caps=1 devices=3 video_nr=7,8,9 card_label=WfRecorder,wfrscreen1,wfrscreen2 debug=1
    wlstreamer --verbose -d 7

Thanks a lot !

**Comment by b12f** [comment]
Since 0.4.0 is now officially released on the AUR, I'll consider this closed.
